brainscore_language/plugins/pereira2018/ceiling_packaging.py

```diff
@@ -0,0 +1,305 @@
+from tqdm import tqdm, trange
+
+from brainio.assemblies import DataAssembly, array_is_element, walk_coords, merge_data_arrays
+from brainscore_core.metrics import Score
+from brainscore_language import load_benchmark
+from brainscore_language.utils import fullname
+from brainscore_language.utils.s3 import upload_data_assembly
+from brainscore_language.utils.transformations import apply_aggregate
+
+_logger = logging.getLogger(__name__)
+
+"""
+Because ceiling estimations for Pereira2018-linear benchmarks are very costly to run (many hours), we pre-compute them,
+and store the resulting ceiling (including per-neuroid estimates) on S3.
+The code in this file was run only once, and is kept here for reference.
+"""
+
+
+def upload_ceiling(experiment):
+    benchmark = load_benchmark(f'Pereira2018.{experiment}-linear')
+    ceiler = ExtrapolationCeiling()
+    ceiling = ceiler(benchmark.data, metric=benchmark.metric)
+    _logger.info(f"Uploading ceiling {ceiling}")
+    # Note that because we cannot serialize complex objects to netCDF, attributes like 'raw' and 'bootstrapped_params'
+    # will get lost during upload
+    upload_data_assembly(ceiling,
+                         assembly_identifier=benchmark.identifier,
+                         assembly_prefix='ceiling_')
+    # also upload raw (and raw.raw) attributes
+    upload_data_assembly(ceiling.raw,
+                         assembly_identifier=benchmark.identifier,
+                         assembly_prefix='ceiling_raw_')
+    upload_data_assembly(ceiling.raw.raw,
+                         assembly_identifier=benchmark.identifier,
+                         assembly_prefix='ceiling_raw_raw_')
+
+
+def v(x, v0, tau0):  # function for ceiling extrapolation
+    return v0 * (1 - np.exp(-x / tau0))
+
+
+class HoldoutSubjectCeiling:
+    def __init__(self, subject_column):
+        self.subject_column = subject_column
+        self._logger = logging.getLogger(fullname(self))
+        self._rng = RandomState(0)
+        self._num_bootstraps = 5
+
+    def __call__(self, assembly, metric):
+        subjects = set(assembly[self.subject_column].values)
+        scores = []
+        iterate_subjects = self._rng.choice(list(subjects), size=self._num_bootstraps)  # use only a subset of subjects
+        for subject in tqdm(iterate_subjects, desc='heldout subject'):
+            try:
+                subject_assembly = assembly[{'neuroid': [subject_value == subject
+                                                         for subject_value in assembly[self.subject_column].values]}]
+                # run subject pool as neural candidate
+                subject_pool = subjects - {subject}
+                pool_assembly = assembly[
+                    {'neuroid': [subject in subject_pool for subject in assembly[self.subject_column].values]}]
+                score = metric(pool_assembly, subject_assembly)
+                # store scores
+                apply_raw = 'raw' in score.attrs and \
```

Comment:
again, here it is unclear for naive users what the notations of 'raw' refer to

Reply:
#42